// js-init/Gruntfile.js
'use strict'

module.exports = function (grunt) {
  grunt.initConfig({
    clean: {
      build: ['docs', 'tmp'],
      tmp: ['tmp']
    },
    copy: {
      src: {
        files: [{
          cwd: 'src',
          dest: 'docs',
          expand: true,
          src: [
            '**/*.css',
            '**/*.js',
            '**/*.svg',
            'favicon.ico'
          ]
        }]
      }
    },
    htmlmin: {
      dist: {
        options: {
          removeComments: true,
          collapseWhitespace: true
        },
        files: [{
          cwd: 'tmp',
          dest: 'docs',
          expand: true,
          src: [
            '**/*.html',
            '!partials/**'
          ]
        }]
      }
    },
    less: {
      development: {
        options: {
          compress: true,
          yuicompress: true,
          optimization: 2
        },
        files: {
          'src/assets/css/main.css': 'src/assets/less/main.less' // destination file: source file
        }
      }
    },
    'string-replace': {
      dist: {
        files: [{
          cwd: 'src',
          dest: 'tmp',
          expand: true,
          src: [
            '**/*.html',
            '!partials/**'
          ]
        }],
        options: {
          replacements: [{
            pattern: /<!-- @import (.*?) -->/ig,
            replacement: function (match, p1) {
              return grunt.file.read('src/' + p1)
            }
          }]
        }
      }
    },
    watch: {
      less: {
        files: ['src/**/*.less'], // files to watch
        tasks: ['less'],
        options: {
          nospawn: true
        }
      }
    }
  })

  grunt.loadNpmTasks('grunt-contrib-clean')
  grunt.loadNpmTasks('grunt-contrib-copy')
  grunt.loadNpmTasks('grunt-contrib-htmlmin')
  grunt.loadNpmTasks('grunt-contrib-less')
  grunt.loadNpmTasks('grunt-contrib-watch')
  grunt.loadNpmTasks('grunt-string-replace')

  grunt.registerTask('build', ['clean:build', 'copy:src', 'string-replace', 'htmlmin', 'clean:tmp'])
  grunt.registerTask('compile', ['less'])
  grunt.registerTask('default', ['less', 'watch'])
}
